daemon/qrexec-daemon.c

```diff
@@ -746,93 +775,269 @@ static void sanitize_name(char * untrusted_s_signed, char *extra_allowed_chars)
-    if (result < 0) {
-         PERROR("error reading from socket");
-         return -1;
-    }
-    else {
-        response[result] = '\0';
-        if (!strncmp(response, "result=allow\n", sizeof("result=allow\n")-1)) {
-            return 0;
-        } else if (!strncmp(response, "result=deny\n", sizeof("result=deny\n")-1)) {
-            return 1;
-        } else {
-            LOG(ERROR, "invalid response: %s", response);
-            return -1;
-        }
+    pid = fork();
+    switch (pid) {
+        case -1:
+            LOG(ERROR, "Could not fork!");
+            _exit(126);
+        case 0:
+            if (close(fds[0]))
+                abort();
+            if (dup2(fds[1], 0) != 0 || dup2(fds[1], 1) != 1) {
+                PERROR("dup2()");
+                _exit(126);
+            }
+            if (close(fds[1]))
+                abort();
+            char remote_domain_id_str[10];
+            int v = snprintf(remote_domain_id_str, sizeof(remote_domain_id_str), "%d",
+                    remote_domain_id);
+            if (v >= 1 && v < (int)sizeof(remote_domain_id_str)) {
+                execl(policy_program,
+                        "qrexec-policy-exec",
+                        "--",
+                        remote_domain_name,
+                        target_domain,
+                        service_name,
+                        NULL);
+                PERROR("execl");
+            } else {
+                PERROR("snprintf");
+            }
+            _exit(126);
+        default:
+            if (close(fds[1]))
+                abort();
+            size_t result_bytes;
+            int status;
+            // this closes the socket
+            char *result = qubes_read_all_to_malloc(daemon_socket, 64, 4096, &result_bytes);
+            do {
+                if (waitpid(pid, &status, 0) != pid) {
+                    PERROR("waitpid");
+                    _exit(126);
+                }
+            } while (!WIFEXITED(status));
+            if (WEXITSTATUS(status) != 0) {
+                LOG(ERROR, "qrexec-policy-exec failed");
+                _exit(126);
+            }
+            if (!parse_policy_response(result, result_bytes, true, user, target, requested_target, autostart))
+                _exit(126);
+            /* do not free result() as user, target, etc point into it */
     }
 }
 
-
 static void handle_execute_service(
         const int remote_domain_id,
         const char *remote_domain_name,
         const char *target_domain,
         const char *service_name,
         const struct service_params *request_id)
 {
-    int i;
-    int result;
     int policy_pending_slot;
     pid_t pid;
-    char remote_domain_id_str[10];
-    sigset_t sigmask;
+
+    switch (target_domain[0]) {
+        case 'A' ... 'Z':
+        case 'a' ... 'z':
+            if (strcmp(target_domain, "dom0") == 0)
+                target_domain = "@adminvm";
```

Comment:
```
Apr 29 01:24:22 dom0 qrexec-policy-daemon[1425]: qrexec: test.AnyvmDeny: test-inst-vm -> @adminvm: denied: no matching rule found
Apr 29 01:24:22 dom0 kernel: qrexec-daemon[17571]: segfault at 0 ip 0000751c2e41362a sp 00007ffc0e519198 error 4 in libc.so.6[751c2e2c8000+156000] likely on CPU 0 (core 0, socket 0)
Apr 29 01:24:22 dom0 kernel: Code: 8b 04 82 29 c8 c3 66 2e 0f 1f 84 00 00 00 00 00 f3 0f 1e fa 89 f1 89 f8 48 83 e1 3f 48 83 e0 3f 83 f9 30 77 3b 83 f8 30 77 36 <f3> 0f 6f 0f f3 0f 6f 16 66 0f ef c0 66 0f 74 c1 66 0f 74 ca 66 0f
```
I'm not sure where the issue is (and whether is it in this line at all), but since you need the requested target transformed by the policy daemon, you can very well drop doing this in qrexec-daemon to simplify the C part and have it done on one place, not two.